design/metastore.md

```diff
@@ -0,0 +1,102 @@
+Trino can work with Hive 2 and 3, while having specific capabilities enabled for version 3.
+Must of the Hive Metastore Thrift interface kept backward compatibility v2.3.x and v3.x interfaces are almost identical, where version 3 adds Catalog and Workload management metadata.
+
+Code generation should be considered. Having generated bride that will transfer each Thrift API call to our OpenAPI. And/or generate code for each entity described in the Thrift interface can reduce the code in order serialize, map and filter the data model.
+
+Thrift API communicate over a socket, the first step will enable non-secure communication over an additional port. A secure communication can be handled by a external TCP load-balancer first, later the proxy can supply this capability out of the box.
+
+
+Hooks - The Thrift interface include a set of API ends with '_environment_context', which enables the caller passing additional information to the hooks. The hooks mechanism in Hive Metastore is a Java interface implementation that the metastore load and calls on specific operations. We will not support this mechanism at first, but we can have a reference implementation of what to expect in case of adding it to our hooks mechanism.
+
+
+### Discovery
+
+Metastore implementation provides a reference implementation we can test and explore. It still needs to be discovered how to extend functionality to enable versioning.
+
+Finding out more about the following items will help us evaluate our design options:
+
+1. Passing lakeFS's reference/branch information from the metastore client. Explore ways to pass repository/branch/ref information from the client to our metastore.
+1. Side by side - very similar to the current work on side-by-side work with lakeFS and the current object-storage, the user like to run the same application with minimum changes, while enabling some of the data to live inside lakeFS. What needs to change or requirements in order to have the same in the metadata level? Do we need to sync information, manage part of the data and enable mapping on the client side? Can we replace the client side implementation? do we need to spin different endpoints to serve different repositories, branches in lakeFS for parts of the metadata? fallback to existing metastore?
+1. Data model diff/merge - describe the level of entities we use for diff/merge. The operation we need to support in order to resolve conflict. How it effects our import/export path?
+1. Authentication with lakeFS - currently lakeFS secure all access by authenticating each request. What will be with metadata?
+1. Data model - best way to model the metadata information into Graveler to enable 3-way diff / merge that will work in optimal way. Enable diff and merge with context, for example, when a column is added we need to know on which table and merge can identify two new partitions to the same table without a conflict.
+
```

Comment:
Sharing an observation, there are two logical milestones I see here - 
1. lakeFS's Versioned Metastore (LVMS?)
2. Metastore clients integration with LVMS

Given this distinction, the discovery items split as follows: 
### lakeFS's Versioned Metastore
1. Data model diff/merge - describe the level of entities we use for diff/merge.
2. Data model - best way to model the metadata information into Graveler to enable 3-way diff / merge that will work in optimal way.

I think that it will be good to work on the items above in order. The level of entities used for diff/merge will affect how we model the information to optimize these operations. 

### Metastore clients integration with LVMS
1. Passing lakeFS's reference/branch information from the metastore client. Explore ways to pass repository/branch/ref information from the client to our metastore.
2. co-exist with other metastores - explore potential ways to achieve that. 
3. Authentication with lakeFS

IMO 3 is independent of 1 & 2

Reply:
I don't understand the emphasis we put on `optimising` diff/merge. How many items are there in the context of a table? Few 100s/1000s at best? Lets think about worst-case scenario here and see if we really need to think about optimisation of a merge/diff..

Reply:
Missed this one - comment about the diff/merge and data model at the end.  Optimizing is not something we need to focus on, just need to think of a data model that will work for us and map the pros and cons in the context of storage and operations we currently support.

@talSofer about the milestones - the first milestone is getting all the known unknowns and decide on the way we like to address the open/discovery items.
I like LMS or lakeMS - will get to naming pretty soon I guess...
Agree we can parallel the client part (although passing ref can be also done by syntax) to the data model and how we going to use lakeFS in implementing our own metastore.
As I see it the authentication we can start simple - do what HMS does - give none and do not open API. But I thought that we already discover what we can pass from the client (branch, ref and etc) so maybe we can use the same technique to pass credentials.
I'll update the metastore-project with the above and add the additional discovery items.